src/shogun/base/unique.h

```diff
@@ -0,0 +1,44 @@
+#ifndef __SG_UNIQUE_H__
+#define __SG_UNIQUE_H__
+
+#ifdef HAVE_CXX11
+
+#include <memory>
+
+namespace shogun
+{
+
+	/** Holds unique pointer that is deleted once this holder is deleted. 
+	 * Its main usage is to hold a pointer to implementation (pimpl idiom):
+	 *
+	 * class Self;
+	 * Unique<Self> self;
+	 *
+	 */
+	template <typename T>
+		class Unique : protected std::unique_ptr<T>
```

Comment:
To restrict all operations but calling a member function of underlying object with ->. Unique pointer has some things like reset(), release(), etc and they can be easily mixed up with reset() of underlying object. E.g. these are different things but easy to confuse: 

```
std::unique_ptr<T> obj;
obj.reset();
obj->reset();
```